### PR TITLE
Suppress Apex deprecation message

### LINKED ISSUE #15070: Suppress Apex AMP deprecation message until Lightning removes its implementation
### Motivation

We currently get this message when running Lightning with the latest apex version:

```
apex.DeprecatedFeatureWarning: apex.amp is deprecated and will be removed by the end of February 2023. Use [PyTorch AMP](https://pytorch.org/docs/stable/amp.html)
```
But Lightning users will be confused about this.

### Pitch

Suppress it and then replace it with the deprecation of #14416 

### Additional context

[This showed up in our CI todayl.
](https://dev.azure.com/Lightning-AI/lightning/_build/results?buildId=105165&view=logs&j=3f274fac-2e11-54ca-487e-194c91f3ae9f&t=fd22e4d0-068c-5651-f467-789f79e7bb7b)

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #15070

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @tchaton @rohitgr7 @carmocca @justusschock @awaelchli @akihironitta

### CODE REVIEW

**Review by otaj — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Comment by awaelchli** [comment]
The previous warning filter did not work, I need to experiment. 
Suggestions welcome

**Comment by otaj** [comment]
@awaelchli the issue is here https://github.com/Lightning-AI/lightning/blob/6f16e46bdb397bf1898e4d7d9050492924177abd/setup.cfg#L36 , since `DeprecatedFeatureWarning` is a subclass of `FutureWarning` (https://github.com/NVIDIA/apex/blob/48bf0df5f7925f71bdeac66c6995ab5c6f2617e3/apex/__init__.py#L58)

The warning is silenced when not run in pytest, but pytest filter takes precedence. I think we can decide to filter this particular inside every test that uses it.

**Comment by awaelchli** [comment]
@otaj Thank you! So the filter works on the user side, that's what we want. Great! 
Will try to suppress it in the test then.

File: tests/tests_pytorch/models/test_amp.py [hunk 1]
```diff
@@ -23,9 +24,26 @@
...
+    import apex
+
+    if not hasattr(apex, "DeprecatedFeatureWarning"):
+        yield
+        return
+
+    with pytest.warns(apex.DeprecatedFeatureWarning, match="apex.amp is deprecated"):
+        yield
+
+
```
Comment (otaj) [inline tests/tests_pytorch/models/test_amp.py:46]: This could be a `pytest.fixture(autouse=True)`

Comment (awaelchli) [reply tests/tests_pytorch/models/test_amp.py:46]: No, because in some tests we need to conditionally activate it based on parameterizations. 
Feel free to commit something if you want, but I think I won't go deeper myself.

File: src/pytorch_lightning/plugins/precision/apex_amp.py [hunk 2]
```diff
@@ -23,8 +24,17 @@
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 
 if _APEX_AVAILABLE:
+    import apex
     from apex import amp
 
+    # TODO: Remove with deprecation of Apex integration #14416
+    if hasattr(apex, "DeprecatedFeatureWarning"):
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/plugins/precision/apex_amp.py:31]: We would of course add our own deprecation in #14416. But we don't want two deprecation messages at the same time. Also, the apex message does not tell Lightning users that they can switch to precision=16 or precision=bf16

File: tests/tests_pytorch/models/test_amp.py [hunk 3]
```diff
@@ -23,9 +24,26 @@
 from lightning_lite.plugins.environments import SLURMEnvironment
 from pytorch_lightning import Trainer
 from pytorch_lightning.demos.boring_classes import BoringModel, RandomDataset
+from pytorch_lightning.utilities.imports import _APEX_AVAILABLE
 from tests_pytorch.helpers.runif import RunIf
 
 
+@contextmanager
+def _suppress_apex_deprecation():
```
Comment (carmocca) [inline tests/tests_pytorch/models/test_amp.py:32]: You can remove all current test changes and just add a marker when runif is set with apex_amp

```diff
diff --git a/tests/tests_pytorch/conftest.py b/tests/tests_pytorch/conftest.py
index a8b54e737..ff1ddec48 100644
--- a/tests/tests_pytorch/conftest.py
+++ b/tests/tests_pytorch/conftest.py
@@ -289,6 +289,11 @@ def pytest_collection_modifyitems(items: List[pytest.Function], config: pytest.C
     for item in items:
         item.add_marker(deprecation_error)
 
+    apex_deprecation = pytest.mark.filterwarnings("ignore::apex.DeprecatedFeatureWarning")
+    for item in items:
+        if any(marker.name == "skipif" and marker.kwargs.get("amp_apex", False) for marker in item.own_markers):
+            item.add_marker(apex_deprecation)
+
 
 def pytest_addoption(parser):
     parser.addoption("--hpus", action="store", type=int, default=1, help="Number of hpus 1-8")
diff --git a/tests/tests_pytorch/helpers/runif.py b/tests/tests_pytorch/helpers/runif.py
index ae09f5b65..d25c246f8 100644
--- a/tests/tests_pytorch/helpers/runif.py
+++ b/tests/tests_pytorch/helpers/runif.py
@@ -153,6 +153,7 @@ class RunIf:
         if amp_apex:
             conditions.append(not _APEX_AVAILABLE)
             reasons.append("NVIDIA Apex")
+            kwargs["amp_apex"] = amp_apex
 
         if bf16_cuda:
             try:
```

I tested it locally and it works

File: src/pytorch_lightning/plugins/precision/apex_amp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/plugins/precision/apex_amp.py:31]: Sure, but we haven't added our deprecation yet. So I would only replace it once we have ours
